doc/source/administrator/optimization.md

```diff
@@ -383,3 +383,141 @@ Choosing the right _limit to guarantee ratio_ ratio is an art, not a science.
 We suggest **starting with a ratio of 2 to 1** and adjusting from there based on whether you run into problems on your hub.
 For example, if the limit is 10GB, start with a guarantee of 5GB.
 Use a service such as Prometheus + Grafana to monitor the memory usage over time, and use this to decide whether to adjust your ratio.
+
+## Explicit memory and CPU allocated to core pods' containers
+
+The Helm chart creates several k8s Pods that typically have one container
+running in them, but sometimes more. In this section you are guided on how to
+explicitly specify how much CPU and memory they are guaranteed via _requests_
+and how much CPU and memory they are limited to via _limits_.
+
+To decide what requests and limits to set, some background knowledge is
+relevant:
+
+1. A [LimitRange](https://kubernetes.io/docs/concepts/policy/limit-range/)
+   resource in a namespace can provide defaults to containers without explicitly
+   set requests and limits.
+
+   Check the namespace where you deploy JupyterHub for such resource by doing
+   `kubectl get limitrange --namespace <k8s-namespace>`.
+
+1. Setting only resource limits _will make_ k8s assume the same in resource
+   requests, but setting only resource requests _will not make_ k8s assume any
+   resource limits.
+
+1. Setting resource _requests_ guarantees a minimal level of resources for the
+   container and therefore some minimal capacity for it to do its job. Setting
+   resource _limits_ close to or equal to the resource requests can be useful to
+   make the container operational performance more consistent.
+
+1. Requested memory is reserved and unavailable for other containers on a node
```

Comment:
Where did you find this info about memory requests being unavailable? I thought if you had two pods that each request 50% and limit 100% of a node's memory, they would each be able to allocate 100% of the node memory as long as it's not at the same time. But if this is accurate, if requests add up to 100%, no pod will be able to exceed its request at all.

My understanding has been that  'request' is really only an input to the scheduler, while 'limit' is the only input to the runtime (OOMKiller).

Reply:
Oh oh hmmmmm, you are probably correct in your understanding.

I've not questioned this understanding much, I think it was formulated by myself based on how I figured it was unreasonable that a pod that have requested a set of memory would have to wait for memory to be freed up when it currently was used used by another container. At the same time, that process could be very quick and i bet the OOMKiller can take action before all memory is depleted as well.

Reply:
Yeah, I think requests are a lot more basic than folks often expect - I think it's purely scheduler bookkeeping, and has no actual effect once pods are scheduled (except possibly in OOMKiller priority when it's time for something to be killed).

Maybe worth just deleting this point then. Then 👍 to merge.

Reply:
I tested with k3d:

<details>
<description>pods.yaml</description>

```yaml
apiVersion: v1
kind: Pod
metadata:
  name: memory-eater
spec:
  containers:
    - name: container
      image: python:3.9-alpine
      env:
        - name: PYTHONUNBUFFERED
          value: "1"
      command:
        - python3
        - -c
        - |
          import os
          import time
          MB = 1024 * 1024
          chunk_size = 100 * MB  # 100 MB
          chunks = []

          while True:
            chunk = bytearray(chunk_size)
            chunks.append(chunk)
            print(f"{len(chunks) * chunk_size // MB }MB")
            time.sleep(5)
      resources:
        requests:
          memory: 900Mi
        limits:
          memory: 1800Mi

---
apiVersion: v1
kind: Pod
metadata:
  name: memory-squatter
spec:
  containers:
    - name: container
      image: python:3.9-alpine
      env:
        - name: PYTHONUNBUFFERED
          value: "1"
      command:
        - sh
        - -c
        - while true; do sleep 30; done
      resources:
        requests:
          memory: 900Mi
        limits:
          memory: 1800Mi
```
</details>

on a node with 2Gi, and the memory eater is allowed to comfortably allocate almost all of the node's memory being killed. The presence or absence of the memory squatter's request for 900M doesn't have any effect on how much the eater is able to allocate before being killed.

If there is a point to be made here, I think it's that oversubscribing CPU results in things being slower than they could be, but usually not catastrophic, while oversubscribing memory results in hangs and crashes. The same actually goes for *Under*-provisioning: low limits on CPU means things might be slow, while low limits on memory means things will keep getting killed.

Reply:
@minrk thank you!!! This is an excellent summary!

Reply:
I relied on this insight and made another pass